homeassistant/components/tesla_wall_connector/sensor.py

```diff
@@ -0,0 +1,141 @@
+"""Sensors for Tesla Wall Connector."""
+from dataclasses import dataclass
+import logging
+
+from homeassistant.components.sensor import (
+    STATE_CLASS_MEASUREMENT,
+    STATE_CLASS_TOTAL_INCREASING,
+    SensorEntity,
+    SensorEntityDescription,
+)
+from homeassistant.const import (
+    DEVICE_CLASS_ENERGY,
+    DEVICE_CLASS_POWER,
+    DEVICE_CLASS_TEMPERATURE,
+    DEVICE_CLASS_VOLTAGE,
+    ELECTRIC_POTENTIAL_VOLT,
+    ENERGY_KILO_WATT_HOUR,
+    ENTITY_CATEGORY_DIAGNOSTIC,
+    FREQUENCY_HERTZ,
+    POWER_KILO_WATT,
+    TEMP_CELSIUS,
+)
+
+from . import (
+    WallConnectorData,
+    WallConnectorEntity,
+    WallConnectorLambdaValueGetterMixin,
+    prefix_entity_name,
+)
+from .const import DOMAIN, WALLCONNECTOR_DATA_LIFETIME, WALLCONNECTOR_DATA_VITALS
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class WallConnectorSensorDescription(
+    SensorEntityDescription, WallConnectorLambdaValueGetterMixin
+):
+    """Sensor entity description with a function pointer for getting sensor value."""
+
+
+WALL_CONNECTOR_SENSORS = [
+    WallConnectorSensorDescription(
+        key="evse_state",
+        name=prefix_entity_name("State"),
+        entity_category=ENTITY_CATEGORY_DIAGNOSTIC,
+        value_fn=lambda data: data[WALLCONNECTOR_DATA_VITALS].evse_state,
+    ),
+    WallConnectorSensorDescription(
+        key="handle_temp_c",
+        name=prefix_entity_name("Handle Temperature"),
+        native_unit_of_measurement=TEMP_CELSIUS,
+        value_fn=lambda data: round(data[WALLCONNECTOR_DATA_VITALS].handle_temp_c, 1),
+        device_class=DEVICE_CLASS_TEMPERATURE,
+        entity_category=ENTITY_CATEGORY_DIAGNOSTIC,
+        state_class=STATE_CLASS_MEASUREMENT,
+    ),
+    WallConnectorSensorDescription(
+        key="grid_v",
+        name=prefix_entity_name("Grid Voltage"),
+        native_unit_of_measurement=ELECTRIC_POTENTIAL_VOLT,
+        value_fn=lambda data: round(data[WALLCONNECTOR_DATA_VITALS].grid_v, 1),
+        device_class=DEVICE_CLASS_VOLTAGE,
+        state_class=STATE_CLASS_MEASUREMENT,
+        entity_category=ENTITY_CATEGORY_DIAGNOSTIC,
+    ),
+    WallConnectorSensorDescription(
+        key="grid_hz",
+        name=prefix_entity_name("Grid Frequency"),
+        native_unit_of_measurement=FREQUENCY_HERTZ,
+        value_fn=lambda data: round(data[WALLCONNECTOR_DATA_VITALS].grid_hz, 3),
+        state_class=STATE_CLASS_MEASUREMENT,
+        entity_category=ENTITY_CATEGORY_DIAGNOSTIC,
+    ),
+    WallConnectorSensorDescription(
+        key="power",
+        name=prefix_entity_name("Power"),
+        native_unit_of_measurement=POWER_KILO_WATT,
+        value_fn=lambda data: round(
+            (
+                (
+                    data[WALLCONNECTOR_DATA_VITALS].currentA_a
+                    * data[WALLCONNECTOR_DATA_VITALS].voltageA_v
```

Comment:
This looks like we calculate the power from the current and voltage. We don't allow calculating state in integrations that integrate devices or services, besides some exceptions for time. Device and service integrations should just report what their API provides while following the entity design of Home Assistant.

Please remove this sensor.

Reply:
I'm not sure if it's documented.